pkg/gatewayserver/io/ws/lbslns/upstream.go

```diff
@@ -482,7 +482,8 @@ func (updf *UplinkDataFrame) FromUplinkMessage(up *ttnpb.UplinkMessage, bandID s
 func (conf TxConfirmation) ToTxAck(ctx context.Context, tokens io.DownlinkTokens, receivedAt time.Time) *ttnpb.TxAcknowledgment {
 	var txAck ttnpb.TxAcknowledgment
 	if msg, _, ok := tokens.Get(uint16(conf.Diid), receivedAt); ok && msg != nil {
-		txAck.DownlinkMessage = msg
+		downMsg := *msg
```

Comment:
Why do we need a copy ? Is not like you can `Get` twice.

Reply:
Because now `Settings` is a pointer, it gets modified [here](https://github.com/TheThingsNetwork/lorawan-stack/blob/b2d03c3ad5a6d068c6c6596c4e1697b93ddb3993/pkg/gatewayserver/io/ws/ws_test.go#L1037).

I could change the test case but we might somewhere down the line mutate this inadvertently in the code somewhere. Better safe than sorry.

Reply:
Personally I'd rather fix the test case, and in another PR/issue make the downlink tokens also erase the downlink on `Get`. 

If we ever mutate the settings somewhere down the line we need to fix that, not do a copy here.

Reply:
The reason for not erasing the token on `Get` is that this is not a push-pop mechanism. For some reason I don't exactly remember, we may expect multiple ACKs with the same `Diid` and so we use retain the Downlink. I haven't seen this happen so we can indeed pop the Downlink on a valid Tx Ack. 
I don't agree however where this copying should be. I feel like doing it at source is a good idea if we know beforehand that there are elements that may be mutated.

Reply:
Copying in `Get` sounds good to me.